src/cultionet/losses/losses.py

```diff
@@ -1,42 +1,156 @@
 import typing as T
+import warnings
 
-from ..models import model_utils
-
+import numpy as np
 import torch
 import torch.nn.functional as F
 from torch_geometric.data import Data
 import torchmetrics
 
+from . import topological
+from ..models import model_utils
+
+
+def one_hot(targets: torch.Tensor, dims: int) -> torch.Tensor:
+    return F.one_hot(targets.contiguous().view(-1), dims).float()
+
 
 class LossPreprocessing(torch.nn.Module):
     def __init__(self, inputs_are_logits: bool, apply_transform: bool):
         super(LossPreprocessing, self).__init__()
 
         self.inputs_are_logits = inputs_are_logits
         self.apply_transform = apply_transform
+        self.sigmoid = torch.nn.Sigmoid()
 
     def forward(
         self,
         inputs: torch.Tensor,
-        targets: torch.Tensor
-    ) -> T.Tuple[torch.Tensor, torch.Tensor]:
+        targets: torch.Tensor = None
+    ) -> T.Tuple[torch.Tensor, T.Union[torch.Tensor, None]]:
+        """Forward pass to transform logits
+
+        If logits are single-dimension then they are transformed by Sigmoid.
+        If logits are multi-dimension then they are transformed by Softmax.
+        """
         if self.inputs_are_logits:
-            if (len(targets.unique()) > inputs.size(1)) or (targets.unique().max()+1 > inputs.size(1)):
-                raise ValueError(
-                    'The targets should be ordered values of equal length to the inputs 2nd dimension.'
-                )
+            if targets is not None:
+                if (len(targets.unique()) > inputs.size(1)) or (targets.unique().max()+1 > inputs.size(1)):
+                    raise ValueError(
+                        'The targets should be ordered values of equal length to the inputs 2nd dimension.'
+                    )
             if self.apply_transform:
-                inputs = F.softmax(inputs, dim=1, dtype=inputs.dtype).clip(0, 1)
-            targets = F.one_hot(
-                targets.contiguous().view(-1), inputs.shape[1]
-            ).float()
+                if inputs.shape[1] == 1:
+                    inputs = self.sigmoid(inputs)
+                else:
+                    inputs = F.softmax(inputs, dim=1, dtype=inputs.dtype)
+
+            inputs = inputs.clip(0, 1)
+            if targets is not None:
+                targets = one_hot(targets, dims=inputs.shape[1])
+
         else:
             inputs = inputs.unsqueeze(1)
             targets = targets.unsqueeze(1)
 
         return inputs, targets
 
 
+class TopologicalLoss(torch.nn.Module):
```

Comment:
This is a new class, but is not in use anywhere.

Reply:
I'm guessing this will be a pain, but this needs a test before we start using it.